### PR TITLE
etcdserver: apply serialized requests outside auth apply lock

### LINKED ISSUE #6010: Investigate serializable range performance regression
Also reported at https://github.com/grpc/grpc-go/issues/771. Before and after grpc dependency update https://github.com/coreos/etcd/pull/5948, there's huge performance drop in serializable range requests:

<br>

https://github.com/coreos/etcd/commit/07ed4da2ffe871444c83be63e508d2adf4c87175:

```
$ benchmark --endpoints=${IP_1},${IP_2},${IP_3} --conns=100 --clients=1000 range aaaaaaaa --consistency=s --total=100000
bench with serializable range
 100000 / 100000 Boooooooooooooooooooooooooooooooooooooooooooooooo! 100.00%1s

Summary:
  Total:    1.0602 secs.
  Slowest:  0.0742 secs.
  Fastest:  0.0003 secs.
  Average:  0.0069 secs.
  Stddev:   0.0061 secs.
  Requests/sec: 94323.9382

Response time histogram:
  0.000 [1] |
  0.008 [65910] |∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.015 [26302] |∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.022 [5257]  |∎∎∎
  0.030 [1173]  |
  0.037 [983]   |
  0.045 [220]   |
  0.052 [59]    |
  0.059 [80]    |
  0.067 [12]    |
  0.074 [3] |

Latency distribution:
  10% in 0.0015 secs.
  25% in 0.0028 secs.
  50% in 0.0052 secs.
  75% in 0.0094 secs.
  90% in 0.0137 secs.
  95% in 0.0177 secs.
  99% in 0.0340 secs.
```

<br>

https://github.com/coreos/etcd/commit/d9a8a326dfe4691cf90d2c4d2674e3602de812d8

```
$ benchmark --endpoints=${IP_1},${IP_2},${IP_3} --conns=100 --clients=1000 range aaaaaaaa --consistency=s --total=100000
bench with serializable range
 100000 / 100000 Boooooooooooooooooooooooooooooooooooooooooooooooo! 100.00%2s

Summary:
  Total:    2.0064 secs.
  Slowest:  0.2937 secs.
  Fastest:  0.0003 secs.
  Average:  0.0196 secs.
  Stddev:   0.0248 secs.
  Requests/sec: 49841.3578

Response time histogram:
  0.000 [1] |
  0.030 [78888] |∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.059 [13889] |∎∎∎∎∎∎∎
  0.088 [5035]  |∎∎
  0.118 [1551]  |
  0.147 [448]   |
  0.176 [126]   |
  0.206 [40]    |
  0.235 [16]    |
  0.264 [4] |
  0.294 [2] |

Latency distribution:
  10% in 0.0004 secs.
  25% in 0.0005 secs.
  50% in 0.0174 secs.
  75% in 0.0267 secs.
  90% in 0.0507 secs.
  95% in 0.0687 secs.
  99% in 0.1079 secs.
```

<br>

https://github.com/coreos/etcd/commit/d8f0ef0e800d86e6343b073c406b7fe0e2cb3940

```
$ benchmark --endpoints=${IP_1},${IP_2},${IP_3} --conns=100 --clients=1000 range aaaaaaaa --consistency=s --total=100000
bench with serializable range
 100000 / 100000 Boooooooooooooooooooooooooooooooooooooooooooooooo! 100.00%2s

Summary:
  Total:    2.0462 secs.
  Slowest:  0.3614 secs.
  Fastest:  0.0003 secs.
  Average:  0.0200 secs.
  Stddev:   0.0255 secs.
  Requests/sec: 48870.5512

Response time histogram:
  0.000 [1] |
  0.036 [80835] |∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.073 [14376] |∎∎∎∎∎∎∎
  0.109 [3760]  |∎
  0.145 [759]   |
  0.181 [194]   |
  0.217 [58]    |
  0.253 [13]    |
  0.289 [2] |
  0.325 [1] |
  0.361 [1] |

Latency distribution:
  10% in 0.0004 secs.
  25% in 0.0005 secs.
  50% in 0.0175 secs.
  75% in 0.0275 secs.
  90% in 0.0527 secs.
  95% in 0.0709 secs.
  99% in 0.1094 secs.
```

### PR DESCRIPTION
**Pull request by heyitsanthony**
Fixes #6010

/cc @mitake @gyuho @xiang90

### CODE REVIEW

**Comment by gyuho** [comment]
lgtm. Tried in the same env. (Go 1.6.3 and this branch)

```
bench with serializable range
 100000 / 100000 Boooooooooooooooooooooooooooooooooooooooooooooooo! 100.00%1s

Summary:
  Total:	1.0936 secs.
  Slowest:	0.0400 secs.
  Fastest:	0.0003 secs.
  Average:	0.0067 secs.
  Stddev:	0.0045 secs.
  Requests/sec:	91444.5515
```

We got the performance back, at least for the issue.

Thanks.

**Comment by gyuho** [comment]
@xiang90 @heyitsanthony Tried with Go 1.7rc3.

```
$ benchmark --endpoints=${IP_1},${IP_2},${IP_3} --conns=100 --clients=1000 range aaaaaaaa --consistency=s --total=100000
bench with serializable range
 100000 / 100000 Boooooooooooooooooooooooooooooooooooooooooooooooo! 100.00%1s

Summary:
  Total:	1.1713 secs.
  Slowest:	0.2102 secs.
  Fastest:	0.0003 secs.
  Average:	0.0100 secs.
  Stddev:	0.0127 secs.
  Requests/sec:	85377.6860

Response time histogram:
  0.000 [1]	|
  0.021 [85462]	|∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.042 [11546]	|∎∎∎∎∎
  0.063 [2301]	|∎
  0.084 [534]	|
  0.105 [120]	|
  0.126 [27]	|
  0.147 [7]	|
  0.168 [0]	|
  0.189 [1]	|
  0.210 [1]	|
```

Slight regression but it's a separate Go issue, we need to investigate more.

File: etcdserver/v3_server.go [hunk 1]
```diff
@@ -551,6 +507,28 @@ func (s *EtcdServer) authInfoFromCtx(ctx context.Context) (*auth.AuthInfo, error
...
+		if err != nil {
+			return err
+		}
+		if ai = authInfo; ai == nil {
+			ai = &auth.AuthInfo{}
+		}
+		if err = chk(ai); err != nil {
+			return err
+		}
+		get()
```
Comment (xiang90) [inline etcdserver/v3_server.go]: this line of code is not easy to understand. probably we need add some comments to the doSerialize func at least.

File: etcdserver/v3_server.go [hunk 2]
```diff
@@ -151,41 +129,19 @@ func (s *EtcdServer) DeleteRange(ctx context.Context, r *pb.DeleteRangeRequest)
...
-			hdr := &pb.RequestHeader{}
-			if authInfo != nil {
-				hdr.Username = authInfo.Username
-				hdr.AuthRevision = authInfo.Revision
-			}
-
-			result = s.applyV3.Apply(&pb.InternalRaftRequest{Header: hdr, Txn: r})
-
-			if result.err != nil {
-				if result.err == auth.ErrAuthOldRevision {
```
Comment (xiang90) [inline etcdserver/v3_server.go]: where is this logic in the new code?

Comment (gyuho) [inline etcdserver/v3_server.go]: Yeah, did we get performance back from having `s.applyV3Base`?

File: etcdserver/v3_server.go (hunk 1 above)
Comment (heyitsanthony) [inline etcdserver/v3_server.go]: added some comments

File: etcdserver/v3_server.go (hunk 2 above)
Comment (heyitsanthony) [inline etcdserver/v3_server.go]: added the ErrAuthOldRevision check

the perf gain is from avoiding the lock in authApplierV3.Apply()

**Comment by mitake** [comment]
@heyitsanthony lgtm, thanks a lot!

File: etcdserver/v3_server.go [hunk 3]
```diff
@@ -151,41 +129,19 @@ func (s *EtcdServer) DeleteRange(ctx context.Context, r *pb.DeleteRangeRequest)
...
-				break
-			}
-
-			if authInfo == nil || authInfo.Revision == s.authStore.Revision() {
-				break
-			}
-
-			// The revision that authorized this request is obsolete.
-			// For avoiding TOCTOU problem, retry of this request is required.
+		var resp *pb.TxnResponse
```
Comment (xiang90) [inline etcdserver/v3_server.go]:
```go
var (
    resp *pb.TxnResponse
    err error
)

File: etcdserver/v3_server.go [hunk 4]
```diff
@@ -551,6 +507,33 @@ func (s *EtcdServer) authInfoFromCtx(ctx context.Context) (*auth.AuthInfo, error
 	return authInfo, nil
 }
 
+// doSerialize handles the auth logic, with permissions checked by "chk", for a serialized request "get".
+func (s *EtcdServer) doSerialize(ctx context.Context, chk func(*auth.AuthInfo) error, get func()) error {
+	var ai *auth.AuthInfo
```
Comment (xiang90) [inline etcdserver/v3_server.go]:
```go
var (
    ai *auth.AuthInfo
    err error
)
```

Comment (xiang90) [inline etcdserver/v3_server.go]: then we ca remove the authinfo in the for loop block?

**Comment by xiang90** [comment]
lgtm

File: etcdserver/v3_server.go (hunk 4 above)
Comment (heyitsanthony) [inline etcdserver/v3_server.go]: I don't think so, the logic is more subtle than that. I'll add some comments.

Comment (xiang90) [inline etcdserver/v3_server.go]: ok. thanks.

File: etcdserver/v3_server.go (hunk 3 above)
Comment (xiang90) [inline etcdserver/v3_server.go]: Just to group declaration together. no strong opinion.

**Pull request merged**